### PR TITLE
Fix #554: Add support for multiple copy layers using multiple assemblies

### LINKED ISSUE #554: Support for multiple COPY layers
We have large folders inside our container and it would be nice if we could put them on different COPY layers. 
This could be archived for example with multiple assembly files, and for each assembly file create a different COPY statement in the dockerfile.

### PR DESCRIPTION
**Pull request by zer0keefie**
Fix #554 

This was largely motivated by the new Spring Boot support for building docker images. I wanted a similar result (separate layers for release dependencies, snapshot dependencies, and the project artifact) without having to switch maven plugins. It looked like the DockerFileBuilder already had enough support for what I needed, and I just had to add support for multiple assemblies.

I _think_ I hit all the points I was aiming for:
- Allow adding multiple <assembly> elements (in an <assemblies> element, because that seemed cleaner), and create a layer for each of those when generating the Dockerfile.
- Preserve the current semantics when a Dockerfile is supplied: require each <assembly> be referenced in a COPY command.
- Don't force a move to the <assemblies> format: <assembly> directly in <build> is still supported, and you can mix the two.
- Assemblies each have to have a unique name, to avoid collisions. I'm not positive that would happen, but I'd rather just avoid it.
- I think I have `docker:watch` working with multiple assemblies.
- As a result of that, JIB mode should work too.
- Updated the docs.

Because I had a specific use case, and I suspect it's pretty common, I added artifact descriptors for dependencies, release dependencies, and snapshot dependencies.

This covers #554, but I made no attempt to avoid the extra posts discussed in #535.

### CODE REVIEW

**Comment by rhuss** [comment]
@zer0keefie Thanks a lot, that looks really awesome! Unfortunately, I'm quite loaded with work (hmm, probably don't have to tell that all the time as this probably always true ;-), and since this PR is not a small one, it might take a bit until I make my way through. At a first look, it looks like that all mandatory parts are included (like docs and tests). 

Going to jump on to this as quick as possible, but probably only on weekends. 

@rohanKanojia if you have some spare cycles it would be awesome if you could add some review comments, too.

**Comment by rhuss** [comment]
Looks like that the latest rebase introduced a NPE in the e2e-tests.

**Comment by rohanKanojia** [comment]
strange, I'll take a look into this possibly tomorrow.

File: src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java
Comment (rohanKanojia) [inline src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java:259]: I think trackArchiver's destFile is no longer being set here after `assemblyArchiver.createArchive(assembly, assemblyName, "track", source, false, null);` call. This seems to be causing NullPointerException in MappingTrackArchiver#getAssemblyFiles() where `getDestFile()` is called. I think explicitly adding destfile should fix this problem:
```suggestion
            File trackArchiverDestFile = assemblyArchiver.createArchive(assembly, assemblyName, "track", source, false, null);
            trackArchivers.get(assemblyName).setDestFile(trackArchiverDestFile);

            return trackArchivers.getAssemblyFiles(mojoParams.getSession(), assemblyName);
```

**Comment by rohanKanojia** [comment]
@zer0keefie : Thanks a lot for your contribution :heart: . This is a really nice feature. I think it would be nice if we had a quickstart for multiple assemblies mode. Could you please add a sample whenever you revisit this? Also, I would appreciate it if you could add an integration test for this so that someone doesn't accidentally break this in the future.

**Comment by rohanKanojia** [comment]
I think there are a few Sonar Code Smells: https://sonarcloud.io/project/issues?id=io.fabric8%3Adocker-maven-plugin&pullRequest=1387&resolved=false&types=CODE_SMELL&q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc

Would appreciate if you could address them too.

File: src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java
Comment (rohanKanojia) [inline src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java:674]: I see you're trying to find assembly names with the same name. Do you think we can use a `Set` here instead? [Set `add()`](https://docs.oracle.com/javase/6/docs/api/java/util/Set.html#add(E)) method returns false if the element already exists in the collection. I haven't tried this myself but this can be simplified like this:
```java
Set<String> assemblyNames = new HashSet<>();
for (AssemblyConfiguration assemblyConfiguration : assemblyConfigurations) {
    boolean wasElementAbsent = assemblyNames.add(assemblyConfiguration.getName());
    if (!wasElementAbsent) {
        log.error("Multiple assemblies use the name %s. Please assign each assembly a unique name.");
        throw new IllegalArgumentException("Assembly names must be unique");
    }
}
```

File: src/test/java/io/fabric8/maven/docker/config/BuildImageConfigurationTest.java
Comment (rohanKanojia) [inline src/test/java/io/fabric8/maven/docker/config/BuildImageConfigurationTest.java:225]: We can use the expected attribute of the `@Test` annotation to declare `IllegalArgumentException` to be thrown anywhere in the annotated test method like this: `@Test(expected = IllegalArgumentException.class)`

**Comment by zer0keefie** [comment]
Thanks for all the feedback! I've incorporated the suggestions (and merged the latest commit from master), plus added a sample that uses multiple assemblies. I think I got most of the Sonar issues cleaned up, but I'll let the Sonar check tell me what I missed.

The Contributing guide recommended squashing the commits for a PR. I find it easier to keep track of what I changed if the incremental changes are separate commits, but I'll rebase and squash them together once everything looks good.

File: src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java
Comment (zer0keefie) [reply src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java]: For as long as I've been using Java, I never looked at the return value of Set.add(). This is much nicer.

File: src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java
Comment (zer0keefie) [reply src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java:263]: Huh, I must be missing something that was getting set up in the dependency injection, because I can't figure out how this was working before. It's definitely working with this change, though.

**Comment by zer0keefie** [comment]
I think that covers everything. I've added the integration test, took care of most of the Sonar issues, and squashed all of the commits into a single one.

The two remaining code smells Sonar flagged are removing the deprecated code (for the single assembly at the top level) and replacing some unit tests with a single parameterized test. There didn't seem to be a good way to make those tests parameterized without having to rewrite a decent chunk of the tests or upgrading to JUnit 5.

**Comment by rohanKanojia** [comment]
@zer0keefie : Hi, Sorry for the late reply. Thanks for taking the time to address my comments. It looks really good. I'll revisit your PR this weekend.

File: it/multi-assembly/pom.xml
Comment (rohanKanojia) [inline it/multi-assembly/pom.xml:17]:
```suggestion
    <description>Docker Maven Plugin Integration Test with Spring Boot with Multiple Layers</description>
```

File: it/multi-assembly/pom.xml
Comment (rohanKanojia) [inline it/multi-assembly/pom.xml:93]: I think you should either remove this(this would be picked from parent pom) or set as `$[project.version}`

File: samples/multi-assembly/pom.xml
Comment (rohanKanojia) [inline samples/multi-assembly/pom.xml:1]: Thanks for adding a sample :+1: . Could you please add a README.md file too regarding build instructions? Something like we have in JIB sample: https://github.com/fabric8io/docker-maven-plugin/tree/master/samples/spring-boot-with-jib

File: src/test/java/io/fabric8/maven/docker/assembly/DockerAssemblyManagerTest.java
Comment (rohanKanojia) [inline src/test/java/io/fabric8/maven/docker/assembly/DockerAssemblyManagerTest.java:11]: nit, we usually avoid wildcard imports. I think this might be due to IDE configuration.

File: src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java
Comment (rohanKanojia) [inline src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java:223]: nit, I think this can be simplified like this:
```java
return Arrays.stream(line)
        .skip(1)
        .filter(component -> !component.startsWith("--"))
        .findFirst()
        .orElse(null);
```

**Review by rohanKanojia — APPROVED** [review]
Thanks a lot :+1: .

File: it/multi-assembly/pom.xml
Comment (zer0keefie) [reply it/multi-assembly/pom.xml]: Good catch. I removed it.

File: samples/multi-assembly/pom.xml
Comment (zer0keefie) [reply samples/multi-assembly/pom.xml:1]: You're welcome! I've added a README as requested.

File: src/test/java/io/fabric8/maven/docker/assembly/DockerAssemblyManagerTest.java
Comment (zer0keefie) [reply src/test/java/io/fabric8/maven/docker/assembly/DockerAssemblyManagerTest.java]: I always forget to go turn that setting back off... I've replaced them with single imports.

File: samples/multi-assembly/pom.xml
Comment (rohanKanojia) [reply samples/multi-assembly/pom.xml:1]: Thanks a lot!

File: src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java
Comment (zer0keefie) [reply src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java]: That's much nicer; I'll do that instead.

**Comment by rohanKanojia** [comment]
@zer0keefie : Thanks for addressing my comments. I've already approved your PR :heavy_check_mark: . Let's wait for Roland to review this(he is an expert on dmp); Hopefully it would be merged and available in next release

**Comment by spmason** [comment]
Any update on this issue? I've been keeping an eye on it for a while now as it should help me to avoid pushing large images during dev testing

File: it/multi-assembly/src/main/java/io/fabric8/maven/sample/springboot/layers/Application.java
Comment (ohbus) [inline it/multi-assembly/src/main/java/io/fabric8/maven/sample/springboot/layers/Application.java:29]: Add an empty line

```suggestion
}

```

File: src/main/resources/assemblies/dependencies.xml
Comment (ohbus) [inline src/main/resources/assemblies/dependencies.xml:24]: Add an empty line

```suggestion
</assembly>

```

File: samples/multi-assembly/src/main/java/io/fabric8/maven/sample/springboot/layers/Application.java
Comment (ohbus) [inline samples/multi-assembly/src/main/java/io/fabric8/maven/sample/springboot/layers/Application.java:29]:
```suggestion
}

```

File: src/main/resources/assemblies/release-dependencies.xml
Comment (ohbus) [inline src/main/resources/assemblies/release-dependencies.xml:16]: Add an extra line

```suggestion
</assembly>

```

File: src/main/resources/assemblies/snapshot-dependencies.xml
Comment (ohbus) [inline src/main/resources/assemblies/snapshot-dependencies.xml:16]: Add an extra line

```suggestion
</assembly>

```

File: src/test/java/io/fabric8/maven/docker/assembly/TrackArchiverCollectionTest.java
Comment (ohbus) [inline src/test/java/io/fabric8/maven/docker/assembly/TrackArchiverCollectionTest.java:77]: Add an extra line

```suggestion
}

```

**Review by ohbus — APPROVED** [review]

**Comment by zer0keefie** [comment]
I swear I have something wrong in my notification settings, so I just noticed there were some comments to address. I pulled in the suggested whitespace fixes.

**Comment by rohanKanojia** [comment]
Hi, Sorry for late reply. I think we should get this in our upcoming release (0.35.0) let's try to get this merged by the end of coming week so that it's available in upcoming release

File: src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java
Comment (rhuss) [reply src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java:265]: Is this fix alread applied ? If not, would be cool to add it before merging.

**Review by rhuss — APPROVED** [review]
Thanks a lot for the pull request for this very valuable feature. Really like it.

Although I haven't dived into detail into the PR, the implementation looks straight forward and also thank you for taking care about all the extras around the code itself (samples, documentation, the property handler that is often forgotten ;-)

I suggest to merge this PR asap when the conflicts are resolved and the @rohanKanojia comments have been addressed. @rohanKanojia feel free to merge when you think its the right time for it.

File: src/main/asciidoc/inc/build/_configuration.adoc
Comment (rhuss) [inline src/main/asciidoc/inc/build/_configuration.adoc:13]: I think the enhanced assembly documentation is missing for the property style configuration. Could you please check and add it if missing there ?

**Comment by rhuss** [comment]
ah yes, and an entry in `docs/changelog.md` would be very cool, too.

**Comment by rhuss** [comment]
fyi, I resolved the conflicts as they seemed to be trivial to resolve.

File: src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java
Comment (rohanKanojia) [reply src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java:292]: I checked and seems like the fix is applied. I'm resolving this comment.

File: src/main/asciidoc/inc/build/_configuration.adoc
Comment (rohanKanojia) [reply src/main/asciidoc/inc/build/_configuration.adoc:13]: Are you talking about configuring assembly via properties? I'm trying this but somehow this isn't working

```
mvn docker:build -Ddocker.imagePropertyConfiguration=override \
  -Ddocker.from=fabric8/java-centos-openjdk8-jdk:1.5.6 \
  -Ddocker.cmd="java -jar /app/dmp-sample-multi-assembly-0.34-SNAPSHOT.jar" \
  -Ddocker.ports.1=8080 \
  -Ddocker.assemblies.1.descriptorRef=dependencies \
  -Ddocker.assemblies.1.name=deps \
  -Ddocker.assemblies.1.targetDir=/app/lib \
  -Ddocker.assemblies.2.descriptorRef=artifact \
  -Ddocker.assemblies.2.targetDir=/app
```

**Comment by rohanKanojia** [comment]
I want to merge this PR and release 0.35.0, I think we can address the assembly property configuration section later on(I think it's some misconfiguration from my side), XMl configuration seems to be working fine.

**Pull request merged**
